pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1459 @@
+	err := t.secretClient.Delete(TokenNamespace, name, options)
+	if err == nil {
+		return nil
+	}
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return apierrors.NewInternalError(fmt.Errorf("failed to delete token %s: %w", name, err))
+}
+
+func (t *Store) get(ctx context.Context, name string, options *metav1.GetOptions) (*ext.Token, error) {
+	sessionID := t.auth.SessionID(ctx)
+
+	// note: have to get token first before we can check for a user mismatch
+	token, err := t.SystemStore.Get(name, sessionID, options)
+	if err != nil {
+		return nil, err
+	}
+
+	user, err := t.auth.UserName(ctx, &t.SystemStore)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return nil, apierrors.NewNotFound(GVR.GroupResource(), name)
+	}
+
+	return token, nil
+}
+
+func (t *SystemStore) Get(name, sessionID string, options *metav1.GetOptions) (*ext.Token, error) {
+	// Core token retrieval from backing secrets
+	// We try to go through the fast cache as much as we can.
+
+	var err error
+	var currentSecret *corev1.Secret
+	empty := metav1.GetOptions{}
+	if options == nil || *options == empty {
+		currentSecret, err = t.secretCache.Get(TokenNamespace, name)
+	} else {
+		currentSecret, err = t.secretClient.Get(TokenNamespace, name, *options)
+	}
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve token %s: %w", name, err))
+	}
+	token, err := tokenFromSecret(currentSecret)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to extract token %s: %w", name, err))
+	}
+
+	token.Status.Current = token.Name == sessionID
+	token.Status.TokenValue = ""
+	return token, nil
+}
+
+func (t *Store) list(ctx context.Context, options *metav1.ListOptions) (*ext.TokenList, error) {
+	user, err := t.auth.UserName(ctx, &t.SystemStore)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+
+	sessionID := t.auth.SessionID(ctx)
+
+	return t.SystemStore.list(user, sessionID, options)
+}
+
+// ListForUser returns the set of token owned by the named user. It is an
+// internal call invoked by other parts of Rancher
+func (t *SystemStore) ListForUser(userName string) (*ext.TokenList, error) {
+	// As internal call this method can use the cache of secrets.
+	// Query the cache, with proper label selector
+	quest := labels.Set(map[string]string{
+		UserIDLabel: userName,
+	}).AsSelector()
+	secrets, err := t.secretCache.List(TokenNamespace, quest)
+
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to list tokens for user %s: %w", userName, err))
+	}
+
+	var tokens []ext.Token
+	for _, secret := range secrets {
+		token, err := tokenFromSecret(secret)
+		// ignore broken tokens
+		if err != nil {
+			continue
+		}
+
+		// Note: We do not care about `Current` in the caller, so no
+		// code here to set that field properly.
+		tokens = append(tokens, *token)
```

Comment:
Note: In most places we set the virtual field `<token>.Status.Current` in the result of `tokenFromSecret` to a proper value. Here, where we know that the caller does not care about the flag we do not bother either.